lib/rules/keyframe-selector-notation/index.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+
+const ruleName = 'keyframe-selector-notation';
+
+const messages = ruleMessages(ruleName, {
+	expected: (selector, fixedSelector) => `Expected "${selector}" to be "${fixedSelector}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/list/keyframe-selector-notation',
+};
+
+const INTERCHANGEABLE_KEYFRAME_SELECTORS = new Set(['0%', '100%', 'from', 'to']);
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary, _, context) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: ['keyword', 'percentage'],
+		});
+
+		if (!validOptions) return;
+
+		const optionFuncs = Object.freeze({
+			keyword: {
+				expFunc: isKeyframeSelectorKeyword,
+				fixFunc: asKeyframeSelectorKeyword,
+			},
+			percentage: {
+				expFunc: isKeyframeSelectorPercentage,
+				fixFunc: asKeyframeSelectorPercentage,
+			},
+		});
```

Comment:
[suggest] IMO, we can simplify the code like this:

```js
const PERCENTAGE_SELECTORS = new Set(['0%', '100%']);
const KEYWORD_SELECTORS = new Set(['from', 'to']);

// ...

/**
 * @type {Record<primary, {
 *   expFunc: (selector: string) => boolean,
 *   fixFunc: (selector: string) => string,
 * }>}
 */
const optionFuncs = Object.freeze({
	keyword: {
		expFunc: (selector) => KEYWORD_SELECTORS.has(selector),
		fixFunc: (selector) => {
			return selector === '0%' ? 'from' : 'to';
		},
	},
	percentage: {
		expFunc: (selector) => PERCENTAGE_SELECTORS.has(selector),
		fixFunc: (selector) => {
			return selector === 'from' ? '0%' : '100%';
		},
	},
});

// ...

if (!KEYWORD_SELECTORS.has(normalizedSelector) && !PERCENTAGE_SELECTORS.has(normalizedSelector)) {
	return;
}
```

But this is just my opinion. The current code is enough, and you can keep it as-is. 👍🏼

Reply:
Definitely agree that this is cleaner in hindsight! I'll look to bring this in.